frame/staking/src/offchain_election.rs

```diff
@@ -155,19 +157,149 @@ pub fn get_balancing_iters<T: Trait>() -> usize {
 	}
 }
 
+/// Find the maximum `len` that a compact can have in order to fit into the block weight.
+///
+/// We start by assuming `voters == size.nominators`, i.e. all nominators are active in the
+/// solution. Then we do a binary search until we find the ideal value.
+///
+/// Thus, this only returns a value between zero and `size.nominators`.
+pub fn maximum_compact_len<W: crate::WeightInfo>(
+	winners_len: u32,
+	size: ElectionSize,
+	max_weight: Weight,
+) -> u32 {
+	use sp_std::cmp::Ordering;
+	let weight_with = |voters: u32| -> Weight {
+		W::submit_solution_better(
+			size.validators.into(),
+			size.nominators.into(),
+			voters,
+			winners_len,
+		)
+	};
+
+	let mut voters = size.nominators.max(2);
+	let mut step = voters;
+	loop {
+		let current = weight_with(voters);
+		step = step / 2;
+		if step == 0 {
+			// We might have reduced less than expected due to rounding error. Reduce one last time.
+			while weight_with(voters) > max_weight {
+				voters -= 1;
+			}
+			// For the opposite case.
+			while weight_with(voters + 1) < max_weight {
+				voters += 1;
+			}
+			break;
+		}
+		match current.cmp(&max_weight) {
+			Ordering::Less => {
+				voters += step;
+			}
+			Ordering::Greater => {
+				voters -= step;
+			}
+			Ordering::Equal => {
+				break;
+			}
+		}
+	}
+
+	voters
+}
+
+/// Greedily reduce the size of the a solution to fit into the block, w.r.t. weight.
+///
+/// The weight of the solution is foremost a function of the number of voters (i.e.
+/// `compact.len()`). Aside from this, the other components of the weight are invariant. The number
+/// of winners shall not be changed (otherwise the solution is invalid) and the `ElectionSize` is
+/// merely a representation of the total number of stakers.
+///
+/// Thus, we reside to stripping away some voters. This means only changing the `compact` struct.
+///
+/// Note that the solution is already computed, and the winners are elected based on the merit of
+/// teh entire stake in the system. Nonetheless, some of the voters will be removed further down the
+/// line.
+///
+/// Indeed, the score must be computed **after** this step. If this step reduces the score too much,
+/// then the solution will be discarded.
+pub fn trim_to_weight<T: Trait, FN>(
+	maximum_allowed_voters: u32,
+	mut compact: CompactAssignments,
+	nominator_index: FN,
+) -> CompactAssignments
+where
+	for<'r> FN: Fn(&'r T::AccountId) -> Option<NominatorIndex>,
+{
+	if let Some(to_remove) = compact.len().checked_sub(maximum_allowed_voters as usize) {
+		// grab all voters and sort them by least stake.
+		let mut voters_sorted = <Nominators<T>>::iter()
+			.map(|(who, _)| {
+				(
+					who.clone(),
+					<Module<T>>::slashable_balance_of_vote_weight(&who),
+				)
+			})
+			.collect::<Vec<_>>();
+		voters_sorted.sort_by_key(|(_, y)| *y);
```

Comment:
That's pretty heavy, isn't there a way to cache it between `trim_to_weight` calls?

Reply:
each OCW will do this only once per era, so sadly no :( 

To be very pedantic, we can remove the read of `Voters<T>::iter()` (cc @shawntabrizi) but this is surely already read in the same OCW process earlier, so it must be cached in the overlay.